feat(STONEINTG-1010): add doc for group snapshot (#147)
konflux-ci/docs: docs/modules/ROOT/pages/how-tos/testing/integration/group-snapshots.adoc

= Creating a group snapshot

In {ProductName}, the integration service provides `group` snapshot to allow users to submit multiple related pull/merge requests under one application to their repos, have them tested together and have the CI results posted back to pull/merge requests.

== Procedures
To have `group` snapshot created for multiple pull/merge requests, complete the following steps:

. Identify the application and components that are related and need to be updated.
. Navigate to their git repos.

> . Create branch with the same branch name for these git repos.

Reviewer: ```suggestion
. Create branches with the same branch name for all these git repos.
```

> . Commit code changes to branch created above and create pull/merge request.

Reviewer: ```suggestion
. Commit code changes to branches created above and create pull/merge requests.
```

Author: Updated. Thanks.

= Creating a group snapshot

> In {ProductName}, the integration service provides `group` snapshot to allow users to submit multiple related pull/merge requests under one application to their repos, have them tested together and have the CI results posted back to pull/merge requests.

Reviewer 2: Does the integration service disable the creation of non-group snapshots?

How does the system handle updating the ITS runs for successive group snapshots (i.e. what happens when group/component1's build finishes the group/component2's build finishes)?

Author: Integration service won't disable non-group snapshot. Component snapshots work as before.
Integration tests running for component snapshot work as before. Group snapshot give users one chance to run some integration test for multiple pull/merge request in one snapshot together. [Here](https://docs.google.com/document/d/1LnS_iv3IuOunFdz9c0VcGd0S7xbVCxsS4mmc-rv3xyo/edit#heading=h.76hl0cvwxpfc) is the spike doc.

> In {ProductName}, the integration service provides extra `group` snapshot to allow users to submit multiple related pull/merge requests under one application to their repos, have them tested together and have the CI results posted back to pull/merge requests, along with the component snapshot created for each component.

Author: Yes

== Procedures
To have `group` snapshot created for multiple pull/merge requests, complete the following steps:

. Identify the application and components that are related and need to be updated.
. Navigate to their git repos.
. Create branches with the same branch name for all of these git repos.
. Commit code changes to branches created above and create pull/merge requests.

> After every build pipelinerun succeeds and component snapshot is created, group snapshot will be created for the new built component container images and other components. And then integration test runs for component and group snapshot.

Reviewer 2: Would something along these lines work? Does it improve clarity?
```suggestion
After every build pipelinerun succeeds and component snapshot is created, a group snapshot will be created containing the latest build for all components which are grouped by the branch names. The configured ItegrationTestScenario(s) will then be run with the group snapshot.
```

Author: Updated. Thanks.

> After every build pipelinerun succeeds and component snapshot is created, a group snapshot will be created containing the new built component container images and other components. The configured ItegrationTestScenario(s) will then be run with the component and group snapshot.

Reviewer 2: The configured ITS(s) are those using the "group" context?

Author: `group` or `application` context ITS will all be run for group snapshot.

Reviewer 2: Should we make that clearer?

Author: Updated. Thanks.

> NOTE: Group snapshot is created only when there are at least two components with open pull/merge request and the same source branch name. 

Reviewer 2: ```suggestion
NOTE: Group snapshots are created only when there are at least two components with open pull/merge request and the same source branch name. 
```

Each of these builds will also have a non-group snapshot created (and will therefore have multiple ITSs run for every commit)? If so, how will the in-git feedback work, will it differentiate the snapshot types?

Author: Each build will have a non-group snapshot created like before. Git feedback is different from for component and group snapshot. Like [this](https://github.com/hongweiliu17/devfile-sample-python-basic-30517/pull/9/checks?check_run_id=31489715766)  for group snapshot and [another](https://github.com/hongweiliu17/devfile-sample-python-basic-30517/pull/9/checks?check_run_id=31489387106) for component snapshot

> The created group snapshot is like below:

Reviewer 2: Is it valid for users to create their own? If not, the level of detail in the CR might not be necessary.

Author: Users can't create their own group snapshot, the info in the annotation `test.appstudio.openshift.io/group-test-info` is requested by users, and then users can get code and run some test in their integration test task. This is why I paste the snapshot sample. Do you any good idea to show this annotation but not the whole snapshot info? Thanks!

[source]
----
apiVersion: appstudio.redhat.com/v1alpha1
kind: Snapshot
metadata:
  name: snapshot-sample
  namespace: ws-sample-tenant
  labels:
    test.appstudio.openshift.io/type: group <1>
    test.appstudio.openshift.io/pr-group-sha: bc3c81ad3d0140f3fe63c61a0b2389cb76663d226fbd28b00fb0d9f050fb21 <2>
  annotations:
    test.appstudio.openshift.io/pr-group: feature1 <3>

>     test.appstudio.openshift.io/group-test-info: [{"namespace":"ws-sample-tenant","component":"component-sample-1","buildPipelineRun":"component-sample-1-on-pull-request-****","snapshot":"application-sample-****","pullRuestNumber":"1","repoUrl":"https://github.com/sample-org/sample-repo1"},{"namespace":"ws-sample-tenant","component":"component-sample-2","buildPipelineRun":"component-sample-2-on-pull-request-****","snapshot":"application-sample-****","pullRuestNumber":"1","repoUrl":"https://github.com/sample-org/sample-repo2"}] <4>

Reviewer 2: Are the astrixes just you excluding information from the sample CR?

Author: Yes, let me show real text. Thanks.

      containerImage: quay.io/redhat-user-workloads/ws-sample-tenant/application-sample/component-sample1@sha256:0db0a473a6abf5c15c424ab07cfbd5c40c06622fe648d4fe6a6b6abc224a0d0c
      source:
        git:
          url: https://github.com/sample-org/sample-repo1
          revision: fa8b89274a61ef0f1c257b7a84c37aa2ec844109
    - name: component-sample-2
      containerImage: quay.io/redhat-user-workloads/ws-sample-tenant/application-sample/component-sample2@sha256:93bc2f55e2d36e50d9915fd89da7d0bfd2499f0d49becf7cd6ff3a25b40d68f6
      source:
        git:
          url: https://github.com/sample-org/sample-repo2
          revision: ff13da0159d2a556c1fc9859a0f19f6b4472c4ad
----
<1> The label `test.appstudio.openshift.io/type: group` that indicates this is a `group` snapshot.
<2> The label `test.appstudio.openshift.io/pr-group-sha: bc3c81ad3d0140f3fe63c61a0b2389cb76663d226fbd28b00fb0d9f050fb21` that indicates the sha256 of pr group derived from branch name and is used to query the k8s resource build pipelinerun and snapshots belonging to the same pr group. Build pipelineruns and snapshots built for the branch with the same branch name have the same pr group.

> <3> The annotation `test.appstudio.openshift.io/pr-group: feature1` that indicates the readable pr group derived from branch name.

Reviewer 2: Is this information used in any CI feedback to the users?

Author: The pr group will be shown on ci feedback. Others won't be used in CI feedback.

> <4> The annotation `test.appstudio.openshift.io/group-test-info` indicates the component snapshots info forming the group snapshot.

Reviewer 2: Is this the entire snapshot that was generated for each of the components included or is it just an annotation representation of the group snapshot?

Author: It contains the component snapshots from the open pull/merge request only

<5> The list of components and their container images. All components updated by pull/merge request with the same pr group will be included. For components which are not updated, the data from Global Candidate List will be used.

== Verification

> As well as integration test result running for component snapshots, integration test result running for `group` snapshot is also posted back to git pull/merge request via checkRun/commitStatus/comment with pr group name in title.

Reviewer 2: ```suggestion
Integration test results from `group` snapshot are posted back to the git pull/merge request in addition to any integration tests results for the component snapshots. These group snapshot results will have the group name in the title.
```
Does this still represent what you were trying to say? Do you think it improves clarity?

> As well as integration test result running for component snapshots, integration test result running for `group` snapshot is also posted back to git pull/merge request via checkRun/commitStatus/comment with pr group name in title like `Konflux Staging / Konflux Staging / application-sample-enterprise-contract / pr group feature-2`.

Reviewer 2: This is in addition to any ITS(s) that are run without the "group" context, right?

I don't think we need to have the full sample PR title, just indicating that the branch-defining group is in the title should be sufficient?

Author: Updated. Thanks

Author: It is for any type of integration test running for group snapshot, to differentiate it from component snapshot.

image::groupsnapshot.png[role="border" alt="Test Report"]

NOTE: For monorepo, since the code changes of multiple components are from the same pull/merge request, the integration test result will be posted back to git pull/merge request only once.

== Troubleshoot

> Sometimes, you can't see integration test result is posted back to pull/merge request, it might be because the group snapshot is not created. You may check by the following steps to debug the reason/failure:

Reviewer 2: Do users have a recourse if the group snapshot isn't created (i.e. can they manually create a group snapshot ... or just a normal snapshot)? Are you just indicating that it isn't created _yet_ or that something might have prevented it from getting created?

Author: They can't manually create group snapshot. I am just indicating that it isn't created yet or that something might have prevented it from getting created.

> NOTE: For monorepo, since the code changes of multiple components are from the same pull/merge request, the integration test result will be posted back to git pull/merge request only once.

Reviewer 2: This is just for the group results, right? Any ITS run against a single component will report back multiple times?

Reviewer 2: Do we have that behavior documented anywhere for a single component's results?

Author: Sorry, I don't quite understand the meaning of `behavior` here.

Reviewer 2: I think what behavior I was referring to was the _generic_ behavior of having a non-group and a group ITS both reporting back to the PR/MR.

Is this statement really accurate though? There will be multiple ITS run:
* One for the component build itself (i.e. application or component context)
* One of each of the group ITS (i.e. after each component in the group has a finished build). In a monorepo, this might still happen out of synchronization as some artifacts may take longer to build than others.

(i.e. does the integration only produce one group snapshot if it sees that there are other builds running within the same group?)

Author: > Is this statement really accurate though? There will be multiple ITS run:
> 
>     * One for the component build itself (i.e. application or component context)
> 
>     * One of each of the group ITS (i.e. after each component in the group has a finished build). In a monorepo, this might still happen out of synchronization as some artifacts may take longer to build than others.

It makes sense. Updated [here](https://github.com/konflux-ci/docs/pull/147/files#diff-597324517f4bd6aa94238f730978b6c585af513ce9ce037340aeea27105d52d2R54-R65). Thanks.

> (i.e. does the integration only produce one group snapshot if it sees that there are other builds running within the same group?)

Yes, there will be only one group snapshot created after all build pipelineruns succeed.
